src/zkscript/bilinear_pairings/model/miller_loop.py

```diff
@@ -65,457 +296,127 @@ def miller_loop(
-                    stack_length_added = 2 * N_ELEMENTS_EVALUATION_OUTPUT
-                    # After this, the stack is: P Q -Q (2T \pm Q) [ev_(l_(T,T))(P) * ev_(l_(2T,\pm Q))(P) * f_i^2]
-                    out += line_eval_times_eval(
-                        take_modulo=False, check_constant=False, clean_constant=False, positive_modulo=False
-                    )
-                    out += Script.parse_string(" ".join(["OP_FROMALTSTACK"] * N_ELEMENTS_MILLER_OUTPUT))
-                    out += line_eval_times_eval_times_miller_loop_output(
-                        take_modulo=take_modulo_F,
-                        positive_modulo=positive_modulo_i,
-                        check_constant=False,
-                        clean_constant=clean_final,
-                        is_constant_reused=False,
-                    )
-
-        # After this, the stack is: (t-1)Q miller(P,Q)
-        out += roll(
-            position=N_ELEMENTS_MILLER_OUTPUT + 3 * N_POINTS_TWIST + N_POINTS_CURVE - 1,
-            n_elements=2 * N_POINTS_TWIST + N_POINTS_CURVE,
-        )
-        out += Script.parse_string(" ".join(["OP_DROP"] * (2 * N_POINTS_TWIST + N_POINTS_CURVE)))
-
-        return optimise_script(out)
+                    out += self.pad_eval_times_eval_times_eval_times_eval_to_miller_output
 
-    def miller_loop_input_data(
-        self, point_p: list[int], point_q: list[int], lambdas_q_exp_miller_loop: list[list[list[int]]]
-    ) -> Script:
-        """Input data required to execute the function `miller_loop`.
-
-        Args:
-            point_p (list[int]): Point `P` at which the Miller loop is evaluated.
-            point_q (list[int]): Point `Q` at which the Miller loop is evaluated.
-            lambdas_q_exp_miller_loop (list[list[list[int]]]): lambdas needed to compute the multiplication wQ. See
-                unrolled_multiplication_input.
+            if i < len(self.exp_miller_loop) - 3:
+                # stack in:  [gradient_(2T), P, Q, T, f_i]
+                # stack out: [gradient_(2T), P, Q, T, f_i^2]
+                out += self.miller_loop_output_square(take_modulo=False, check_constant=False, clean_constant=False)
 
-        Returns:
-            Script pushing the input data required to execute the function `miller_loop`.
-        """
-        out = nums_to_script([self.MODULUS])
-        for i in range(len(lambdas_q_exp_miller_loop) - 1, -1, -1):
-            for j in range(len(lambdas_q_exp_miller_loop[i]) - 1, -1, -1):
-                out += nums_to_script(lambdas_q_exp_miller_loop[i][j])
+            if self.exp_miller_loop[i] == 0:
+                # stack in:  [gradient_(2T), P, Q, T, f_i^2]
+                # stack out: [P, Q, 2T, (f_i^2 * ev_(l_(T,T))(P))]
+                out += self.__one_step_without_addition(
+                    i=i,
+                    take_modulo=[take_modulo_miller_loop_output, take_modulo_point_multiplication],
+                    positive_modulo=positive_modulo_i,
```

Comment:
I would either have a variable clean_constant_i = ... or change the behaviour of positive modulo so they are consistent with each other